data_encryption/models/encrypted_data.py

```diff
@@ -0,0 +1,139 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import json
+import logging
+
+from odoo import api, fields, models
+from odoo.exceptions import AccessError, ValidationError
+from odoo.tools import ormcache
+from odoo.tools.config import config
+from odoo.tools.translate import _
+
+_logger = logging.getLogger(__name__)
+
+try:
+    from cryptography.fernet import Fernet, InvalidToken
+except ImportError as err:  # pragma: no cover
+    _logger.debug(err)
+
+
+class EncryptedData(models.Model):
+    """Model to store encrypted data by environment (prod, preprod...)"""
+
+    _name = "encrypted.data"
+    _description = "Store any encrypted data by environment"
+
+    name = fields.Char(required=True, readonly=True, index=True, help="Technical name")
+    environment = fields.Char(
+        required=True, index=True, help="Concerned Odoo environment (prod, preprod...)",
+    )
+    encrypted_data = fields.Binary(attachment=False)
+
+    _sql_constraints = [
+        (
+            "name_environment_uniq",
+            "unique (name, environment)",
+            "You can not store multiple encrypted data for the same record and \
+          environment",
+        )
+    ]
+
+    def _decrypt_data(self, env):
+        self.ensure_one()
+        cipher = self._get_cipher(env)
+        try:
+            return cipher.decrypt(self.encrypted_data).decode()
+        except InvalidToken:
+            raise ValidationError(
+                _(
+                    "Password has been encrypted with a different "
+                    "key. Unless you can recover the previous key, "
+                    "this password is unreadable."
+                )
+            )
+
+    @api.model
+    @ormcache("self._uid", "name", "env")
+    def _encrypted_get(self, name, env=None):
+        if self.env.context.get("bin_size"):
+            self = self.with_context(bin_size=False)
+        if not self.env.user._is_superuser():
+            raise AccessError(_("Encrypted data can only be read as superuser"))
```

Comment:
@florian-dacosta Why the encryption model is meant to be used only as superuser? It's problematic in my project.

Reply:
Hi @ThomasBinsfeld 
The encryption model is meant to contained all encrypted data (from any model).
So, it may containing many different sensible data (passwords...).

To be secure, we restrict its access to the superuser. Indeed, if one use can access this model, then it can access all sensitive data across all models.
Each module submodule implementing the encryption/decryption of data should manage its case using sudo.

It is the case for data_encryption_server_environement, which handle the encryption of data for all server dependent fields.
https://github.com/OCA/server-env/blob/12.0/server_environment_data_encryption/models/server_env_mixin.py#L22
https://github.com/OCA/server-env/blob/12.0/server_environment_data_encryption/models/server_env_mixin.py#L40

If you create another sub-module made to encrypt/decrypt another type of data, I guess you should implement it with sudo also.
I think it would be a security breach to change this


Why is it problematic for your project?

Reply:
Thanks for your response. I think i understand what my problem is.

In v13, `sudo()` doesn't change the current user. So `self.env.user._is_superuser()` is False, even if the method is called with `sudo()`. That's why i have access errors everywhere in my project.

In `data_encryption`, i need to replace `if not self.env.user._is_superuser():` by `if not self.env.user._is_superuser() and not self.env.su:` if i want to detect the `sudo()`.

In order to preserve the initial behavior, do you think i should replace the two `sudo()` you pointed me by a `with_user(SUPERUSER_ID)`?

Reply:
Mmmh, I do not think it is necessary to keep the super user stuff.

Like you said, sudo worked differently in v12- and the user was changed to super user, which is not the case anymore.
So I guess in v13, we just have to check self.env.su and we do not need anymore to check the user.
I did not work for v13 yet, so I am not so confortable with this and could be mistaken, but I think it is enough.

What do you think?

Reply:
Done :ok_hand: